### ISSUE TITLE
Convenient interface to evaluate and sample from the predictive distribution

### ISSUE THREAD

**Issue by scheidan** [PriorLabs/TabPFN#871]
### Describe the workflow you want to enable

We were running a workshop on tabPFN in our institute and the possibility to learn the complete predictive distribution sparked a lot of interest.

The usage, however, is rather convoluted (see [PRI-54](https://linear.app/priorlabs/issue/PRI-54/document-bar-distribution-bimodal-full-uncertainty-regression-outputs)).<br>It would be great to have a convenient interface to evaluate (and sample) the probability density of the predictive distribution:

$p(y \\mid X\_\\text{new}, X\_\\text{train}, y\_\\text{train})$

### Describe your proposed solution

I've tried to write two helpers. However, the computed densities are somehow off (see demo below)

```python
def predict_pdf(model, X_pred, y):
    """
    Evaluates density p(y | X_pred, X_train, y_train) for a vector of y values.
    Output shape: (X_pred.shape[0], y.size)
    """
    pp = model.predict(X_pred, output_type="full")
    y = np.asarray(y)

    y_eval = torch.tensor(
        np.repeat(y[None, :], X_pred.shape[0], axis=0),
        dtype=pp["logits"].dtype,
        device=pp["logits"].device,
    )
    # Note, before TabPFN v7.1.0 pdf() gave the inverted pdf
    dens = [pp["criterion"].pdf(pp["logits"], y_eval[:, k, None]).cpu().numpy() for k in range(y.shape[0])]
    return np.stack(dens).T


def predict_pdf_samples(model, X_pred, n_samples=1):
    """
    Samples from p(y | X_pred, X_train, y_train).
    Output shape: (X_pred.shape[0], n_samples)
    """
    pp = model.predict(X_pred, output_type="full")

    logits = pp["logits"].repeat_interleave(n_samples, dim=0)
    samples = pp["criterion"].sample(logits)
    return samples.reshape(X_pred.shape[0], n_samples)
```

### Describe alternatives you've considered, if relevant

The density does not match the histogram of the samples. The plot was produced with the code below (`tabpfn v7.1.0`).

<img src="https://github.com/user-attachments/assets/5435a49c-a201-4953-bb56-3fbe24a4a722 " alt="Image" width="777" data-linear-height="494" />

```python
import numpy as np
from numpy.random import randn, choice
import matplotlib.pyplot as plt
import torch
from tabpfn import TabPFNRegressor

# convenience functions

def predict_pdf(model, X_pred, y):
    """
    Evaluates density p(y | X_pred, X_train, y_train) for a vector of y values.
    Output shape: (X_pred.shape[0], y.size)
    """
    pp = model.predict(X_pred, output_type="full")
    y = np.asarray(y)

    y_eval = torch.tensor(
        np.repeat(y[None, :], X_pred.shape[0], axis=0),
        dtype=pp["logits"].dtype,
        device=pp["logits"].device,
    )
    # Note, before TabPFN v7.1.0 pdf() gave the inverted pdf
    dens = [pp["criterion"].pdf(pp["logits"], y_eval[:, k, None]).cpu().numpy() for k in range(y.shape[0])]
    return np.stack(dens).T


def predict_pdf_samples(model, X_pred, n_samples=1):
    """
    Samples from p(y | X_pred, X_train, y_train).
    Output shape: (X_pred.shape[0], n_samples)
    """
    pp = model.predict(X_pred, output_type="full")

    logits = pp["logits"].repeat_interleave(n_samples, dim=0)
    samples = pp["criterion"].sample(logits)
    return samples.reshape(X_pred.shape[0], n_samples)


# --- test

# 1) Generate data
n = 180
x = np.random.uniform(1, 10, n)
y = np.array(
    [choice([0.6 * xi + 0.5 * randn(), 0.2 * xi + 0.3 * randn()]) + 0.6 * xi for xi in x]
)
x = x.reshape(-1, 1)

# 2) fit model
mod_tab = TabPFNRegressor()
mod_tab.fit(x, y)

# 3) plot the histogram samples and the pdf in the same plot
#     p(y | xpred=5, xtrain yrain)

x_pred = np.array([[5.0]])
y_grid = np.linspace(np.min(y) - 0.5, np.max(y) + 0.5, 400)

pdf_vals    = predict_pdf(mod_tab, X_pred=x_pred, y=y_grid)[0]
sample_vals = predict_pdf_samples(mod_tab, X_pred=x_pred, n_samples=500)[0]

plt.figure(figsize=(8, 5))
plt.hist(
    sample_vals,
    bins=30,
    density=True,
    alpha=0.45,
    color="steelblue",
    edgecolor="white",
    label="samples",
)
plt.plot(y_grid, pdf_vals, color="darkred", linewidth=2, label="predictive pdf")
plt.xlabel("y")
plt.ylabel("density")
plt.title(f"Predictive distribution at x = {x_pred[0, 0]:.1f}")
plt.legend()
plt.tight_layout()
plt.show()
```

### Additional context

*No response*

### Impact

None

**Comment by jonaslandsgesell** [comment]
Hello Andreas,

I thought you might find [ScoringBench](http://scoringbench.bolt.host/) relevant to your work. It is an evaluation framework we have developed for **distributional regression models** (such as TabPFN, TabICL, and adapted XGBoost variants). We designed it with minimalistic wrappers to make extending it to new models highly straightforward.

* **GitHub Repository:** [jonaslandsgesell/ScoringBench](https://github.com/jonaslandsgesell/ScoringBench)
* **Live Leaderboard:** [ScoringBench Leaderboard](http://scoringbench.bolt.host/)

### Why move beyond point estimates?
While tabular foundation models naturally produce predictive distributions, most current benchmarks still lean on metrics like RMSE or R2. These metrics often reward models that accurately capture the conditional mean while ignoring the actual quality of the predicted distribution.

In our accompanying research, we argue for a necessary shift toward **Proper Scoring Rules**:

* **The Research:** 
    * [Distributional Regression with Tabular Foundation Models](https://arxiv.org/abs/2603.08206)
    * [ScoringBench: A Benchmark for Evaluating Tabular Foundation Models](https://arxiv.org/abs/2603.29928)
* **Key Findings:** We have integrated metrics such as **CRPS**, **CRLS**, and **Energy Score** to provide a richer picture of performance, especially in the "tails" where risk is most concentrated. Our results demonstrate that model rankings shift significantly depending on the scoring rule used, suggesting that training objectives should be as domain-specific as the data itself.
* **Impact of Fine-tuning:** We found that fine-tuning **realTabPFNv2.5** with specific scoring rules (like the Energy Score) yields consistent improvements. This confirms that the training loss shapes these models far beyond what propriety alone guarantees.

ScoringBench includes a live leaderboard to track how different models handle these probabilistic forecasts. If you are working with high-stakes tabular data where asymmetric risk profiles matter, I would love to hear your thoughts on this approach.

Kind regards,

Jonas

**Comment by scheidan** [comment]
Thanks Jonas, ScoringBench looks interesting, I'll have a closer look!

After testing a bit more, it seem to be a regression of `ModelVersion.V2_6`; it works well with `ModelVersion.V2_5`:

<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/3c4ddd3b-e7d2-4f8c-81eb-10d3702b67cd" />

See the slightly updated code below.

<details>

<summary>Updated code</summary>

```python
import numpy as np
from numpy.random import randn, choice
import matplotlib.pyplot as plt
import torch
from tabpfn import TabPFNRegressor

from tabpfn.constants import ModelVersion

# convinence functions

def predict_pdf(model, X_pred, y):
    """
    Evaluates density p(y | X_pred, X_train, y_train) for a vector of y values.
    Output shape: (X_pred.shape[0], y.size)
    """
    pp = model.predict(X_pred, output_type="full")
    y = np.asarray(y)

    y_eval = torch.tensor(
        np.repeat(y[None, :], X_pred.shape[0], axis=0),
        dtype=pp["logits"].dtype,
        device=pp["logits"].device,
    )
    # Note, before TabPFN v7.1.0 pdf() gave the inverted pdf
    dens = [pp["criterion"].pdf(pp["logits"], y_eval[:, k, None]).cpu().numpy() for k in range(y.shape[0])]
    return (np.stack(dens)).T


def predict_pdf_samples(model, X_pred, n_samples=1):
    """
    Samples from p(y | X_pred, X_train, y_train).
    Output shape: (X_pred.shape[0], n_samples)
    """
    pp = model.predict(X_pred, output_type="full")

    logits = pp["logits"].repeat_interleave(n_samples, dim=0)
    samples = pp["criterion"].sample(logits)
    return samples.reshape(X_pred.shape[0], n_samples)


# --- test

# 1) Generate data
n = 180
x = np.random.uniform(1, 10, n)
y = np.array(
    [choice([0.6 * xi + 0.5 * randn(), 0.2 * xi + 0.3 * randn()]) + 0.6 * xi for xi in x]
)
x = x.reshape(-1, 1)

# 2) fit model
mod_2_5 = TabPFNRegressor.create_default_for_version(ModelVersion.V2_5)
mod_2_6 = TabPFNRegressor.create_default_for_version(ModelVersion.V2_6)

mod_2_5.fit(x, y)
mod_2_6.fit(x, y)

# 3) plot the histogram samples and the pdf in the same plot
#     p(y | xpred=5, xtrain yrain)

x_pred = np.array([[5.0]])
y_grid = np.linspace(np.min(y) - 0.5, np.max(y) + 0.5, 400)

pdf_vals_2_5    = predict_pdf(mod_2_5, X_pred=x_pred, y=y_grid)[0]
sample_vals_2_5 = predict_pdf_samples(mod_2_5, X_pred=x_pred, n_samples=500)[0]

pdf_vals_2_6    = predict_pdf(mod_2_6, X_pred=x_pred, y=y_grid)[0]
sample_vals_2_6 = predict_pdf_samples(mod_2_6, X_pred=x_pred, n_samples=500)[0]


fig, axs = plt.subplots(2, 1)
axs[0].hist(
    sample_vals_2_5,
    bins=30,
    density=True,
    alpha=0.45,
    color="steelblue",
    edgecolor="white",
    label="samples",
)
axs[0].plot(y_grid, pdf_vals_2_5, color="darkred", linewidth=2, label="predictive pdf")
axs[0].set_xlabel("y")
axs[0].set_ylabel("density")
axs[0].set_title(f"(ModelVersion.V2_5, tabpfn v7.1.1) Predictive distribution at x = {x_pred[0, 0]:.1f}")
axs[0].legend()

axs[1].hist(
    sample_vals_2_6,
    bins=30,
    density=True,
    alpha=0.45,
    color="steelblue",
    edgecolor="white",
    label="samples",
)
axs[1].plot(y_grid, pdf_vals_2_6, color="darkred", linewidth=2, label="predictive pdf")
axs[1].set_xlabel("y")
axs[1].set_ylabel("density")
axs[1].set_title(f"(ModelVersion.V2_6, tabpfn v7.1.1) Predictive distribution at x = {x_pred[0, 0]:.1f}")
axs[1].legend()

plt.tight_layout()
plt.show()
```

</details>

**Comment by jonaslandsgesell** [comment]
On scoring bench https://scoringbench.bolt.host/ we also observed a degradation of v2.6 compared to v2.5. especially, in the CRPS, the log score, the beta energy family and the continuous Brier score.

This is why we also just now opened an issue https://github.com/PriorLabs/TabPFN/issues/873 

We hope that ScoringBench and the accompanying publications https://arxiv.org/abs/2603.08206 https://arxiv.org/abs/2603.29928 will help to bring attention to distributional forecasting and the application of optimum score estimation https://sites.stat.washington.edu/raftery/Research/PDF/[Gneiting2007jasa.pdf](https://sites.stat.washington.edu/raftery/Research/PDF/Gneiting2007jasa.pdf).

**Comment by jonaslandsgesell** [comment]
@scheidan I added an example ipynb to illustrate how the scoringbench repo can serve as a representation normalization layer which also evaluates measures of "probabilistic goodness of fit" (in terms of proper scoring rules)

https://github.com/jonaslandsgesell/ScoringBench/blob/master/Comparison.ipynb

**Comment by bhvieira** [comment]
My own explorations indicate this might still occur with V3.0

**Comment by bejaeger** [comment]
Thanks a lot for looking at this and reporting!
We're currently working on better documentation and usage of the predictive distribution. Your input is very valuable.

**Comment by eliott-kalfon** [comment]
Hi @scheidan, we have since then worked on visualising this predictive distributions. You may find the following useful:

- Our new docs [page](https://docs.priorlabs.ai/capabilities/predictive-distribution) on the topic
- An example [script](https://github.com/PriorLabs/TabPFN/blob/main/examples/plot_regression_distribution.py) on the TabPFN package
- An example [script and README](https://github.com/PriorLabs/tabpfn-extensions/tree/main/examples/predictive_distribution) on the TabPFN extensions

Let me know if you have more questions. Thank you for raising this issue, this was a good excuse to build a better interface for TabPFN. Have a nice day :)

**Comment by eliott-kalfon** [comment]
I am marking this issue as closed in the meantime

**Comment by scheidan** [comment]
Thanks! The new documentation is a big improvement in my view! 

Some small confusions remain:

- Is the support of the predictive distributions bounded or unbounded (see Fig. 2 b-c of Müller et al. 2021)?
- I assume $z$ are the logits in [this](https://docs.priorlabs.ai/capabilities/predictive-distribution#what-is-the-bar-distribution) equation. Would be nice to spell that out.


Müller, S., Hollmann, N., Arango, S.P., Grabocka, J., Hutter, F., 2021. Transformers Can Do Bayesian Inference. Presented at the International Conference on Learning Representations.

**Comment by eliott-kalfon** [comment]
Hi @scheidan, good point, just opened a docs PR to make the change. The support of the distribution is unbounded (see [here](https://github.com/PriorLabs/docs/pull/75)). We kept the docs simple. Cheers :)

> Thanks! The new documentation is a big improvement in my view!
> 
> Some small confusions remain:
> 
> * Is the support of the predictive distributions bounded or unbounded (see Fig. 2 b-c of Müller et al. 2021)?
> * I assume 
>     z
>    are the logits in [this](https://docs.priorlabs.ai/capabilities/predictive-distribution#what-is-the-bar-distribution) equation. Would be nice to spell that out.
> 
> Müller, S., Hollmann, N., Arango, S.P., Grabocka, J., Hutter, F., 2021. Transformers Can Do Bayesian Inference. Presented at the International Conference on Learning Representations.
